File: BudgetApp.Backend/BudgetApp.Backend.Dto/Category.cs
using System;
using System.Runtime.Serialization;
using System.Text.Json.Serialization;
using BudgetApp.Backend.Dto.Interfaces;
using MongoDB.Bson.Serialization.Attributes;

namespace BudgetApp.Backend.Dto
{
    [ViewFriendlyName("Categories")]
    [ViewShown(true)]
    [TransferableDataType]
    [DataSchema]
    [ViewForceReload(true)]
    public class Category : DataTransferObject
    {
        [BsonElement("name")]
        [JsonPropertyName("name")]
        [DataFriendlyName("Name")]
        [DataPlaceholder("e.g. General")]
        [DataType(DataTypeEnum.String)]
        [DataRelationshipView(true)]
        [DataRequired]
        public string Name { get; set; }

        [BsonElement("color")]
        [JsonPropertyName("color")]
        [DataFriendlyName("Color")]
        [DataType(DataTypeEnum.Color)]
        [DataRelationshipViewColor(true)]
        [DataRequired]
        public string Color { get; set; }

        [BsonElement("threshold")]
        [JsonPropertyName("threshold")]
        [DataFriendlyName("Threshold")]
        [DataPlaceholder("e.g 500.00")]
        [DataType(DataTypeEnum.Number)]
        [DataRequired]
        [DataSensitive]
        public decimal Threshold { get; set; }

        // [BsonElement("togglePicker")]
        // [JsonPropertyName("togglePicker")]
        // [DataHidden]
        // [DataType(DataTypeEnum.Boolean)]
        // public bool TogglePicker { get; set; }

        // [BsonElement("isDarkColor")]
        // [JsonPropertyName("isDarkColor")]
        // [DataHidden]
        // [DataType(DataTypeEnum.Boolean)]
        // public bool IsDarkColor { get; set; }

        public override bool ValidateInsert(params string[] args)
        {
            throw new NotImplementedException();
        }

        public override bool ValidateUpdate(params string[] args)
        {
            throw new NotImplementedException();
        }

        public override bool ValidateDelete(params string[] args)
        {
            throw new NotImplementedException();
        }
    }
}

File: budget-app-fe/src/app/classes/dto/category.ts
import {IDataTransferObject} from './interfaces/data-transfer-object.interface';
import {INamedDto} from './interfaces/expense-dto.interface';

export interface IColor {
  color: string;
  isDarkColor: boolean;
}

export interface IColorDto extends INamedDto, IColor {

}

export class Category implements IColorDto {
  sanitize(): void {
    this.threshold = +this.threshold;
  }
  public _id: string;
  public name: string = '';
  public color: string = '#d6d2d2';
  public threshold: number = 0;
  public togglePicker: boolean = false;
  public isDarkColor: boolean = false;
  removable: boolean = true;
  isDirty: boolean;
  isDeleted: boolean;

}

